### DISCUSSION TITLE
How to deal with OPTIONS requests in BFF?

### DISCUSSION THREAD

**Discussion by ligiaandreica** [remix-run/remix#5067] [Q&A] [answered]
I am using remix as BFF, implementation based on [docs](https://remix.run/docs/en/v1/guides/bff). `The client` (on a different domain) calls `BFF` to get some data and then the requests go to `downstream APIs`. Basically, it's the preflight OPTIONS request to verify the client can do the GET.

This was working fine with some `remix-utils` `cors` setup until `v1.8.0` of remix.

Started getting `Error: Invalid request method "OPTIONS"` once I've upgraded from `v1.7.6`. This error seems expected according to the [release docs](https://github.com/remix-run/remix/releases/tag/remix%401.8.0).
```
ErrorResponse {
  status: 405,
  statusText: 'Method Not Allowed',
  internal: true,
  data: 'Error: Invalid request method "OPTIONS"',
  error: Error: Invalid request method "OPTIONS" ...
}
```

Am I missing some configuration or is this a new limitation?
@jacob-ebey or @brophdawg11 could you please advise as this is part of https://github.com/remix-run/remix/pull/4385?

**Comment by brophdawg11** [comment]
Hm - this wasn't an intentional _new_ restriction, but rather one that [already existed in Remix 1.7.6](https://github.com/remix-run/remix/blob/remix@1.7.6/packages/remix-server-runtime/server.ts#L686).  We even have an [integration test](https://github.com/remix-run/remix/blob/remix%401.7.6/integration/catch-boundary-test.ts#L253) asserting this in 1.7.6.

Any chance you could put together a working reproduction of the issue so we can dig in further?

**Comment by ligiaandreica** [comment, marked as answer]
Sure, [here](https://github.com/ligiaandreica/remix-upgrade) it is.

[client](https://github.com/ligiaandreica/remix-upgrade/tree/main/client) CRA app with react-query to force preflight `OPTIONS` requests.
 - `npm install`, `npm run start` -> on port `3001`
 - data logged in the console

Identical `api.data.ts` in both remix apps to serve some mock data:

[remix-1.7.6](https://github.com/ligiaandreica/remix-upgrade/tree/main/remix-1.7.6) created via `npx create-remix`, `Just the basic`, `Express server` + `remix-utils` package
-  [api.data.ts](https://github.com/ligiaandreica/remix-upgrade/blob/main/remix-1.7.6/app/routes/api.data.ts) responds with some mocked data to the client
<img width="763" alt="image" src="https://user-images.githubusercontent.com/59609285/212114686-a1d402ac-c356-411a-b673-62b950c45395.png">

![image](https://user-images.githubusercontent.com/59609285/212115635-dc606794-4263-4806-9532-a9a041110445.png)


[remix-1.8.0](https://github.com/ligiaandreica/remix-upgrade/tree/main/remix-1.8.0) created via `npx create-remix`, `Just the basic`, `Express server` + `remix-utils` package
-  [api.data.ts](https://github.com/ligiaandreica/remix-upgrade/blob/main/remix-1.8.0/app/routes/api.data.ts) responds with some mocked data to the client
<img width="1510" alt="image" src="https://user-images.githubusercontent.com/59609285/212115238-509e3a80-f87d-44df-b265-c19a37eec0c5.png">

![image](https://user-images.githubusercontent.com/59609285/212115175-bbb88763-a60d-49be-b330-298875d731b1.png)

**Reply by brophdawg11** [reply]
ah ok thank you - it looks like Remix 1.7.6 only validated the method on document and data requests, but not resource requests.  I'm going to convert this to an issue so we can look into that.  Thanks!

**Reply by brophdawg11** [reply]
New Issue: https://github.com/remix-run/remix/issues/5093

**Comment by ligiaandreica** [comment]
This seems to be the [check](https://github.com/sergiodxa/remix-utils/blob/main/src/server/cors.ts#L63)  done by `remix-utils` `cors`. Wondering if the new error on `OPTIONS` might be coming from [remix-run/react-router](https://github.com/remix-run/react-router/blob/03b7f408c36f722599df14a1bfab29dabc35560d/packages/router/router.ts#L2073).
